File: typescript/numbers/src/numbers_iterative.ts
export function fibonacci(n: number): number {
  return fibonacci_iter(n, 0, 1);
}

function fibonacci_iter(n: number, acc2: number, acc1: number): number {
  if (n <= 0) return acc2;
  if (n <= 2) return acc1 + acc2;
  return fibonacci_iter(n - 1, acc1, acc1 + acc2);
}

export function factorial(n: number): number {
  return factorial_iter(n, 1);
}

function factorial_iter(n: number, acc: number): number {
  if (n <= 1) return acc;
  return factorial_iter(n - 1, n * acc);
}

export function sum_numbers(n: number): number {
  return sum_numbers_iter(n, 0);
}

function sum_numbers_iter(n: number, acc: number): number {
  if (n <= 0) return acc;
  return sum_numbers_iter(n - 1, n + acc);
}

File: elixir/numbers/lib/numbers_iterative.ex
defmodule NumbersIterative do
  def fibonacci(n), do: fibonacci_iter(n, 0, 1)

  defp fibonacci_iter(n, acc2, acc1) when n <= 1, do: n
  defp fibonacci_iter(n, acc2, acc1) when n == 2, do: acc1 + acc2
  defp fibonacci_iter(n, acc2, acc1), do: fibonacci_iter(n - 1, acc1, acc1 + acc2)

  def factorial(n), do: factorial_iter(n, 1)

  defp factorial_iter(0, accumulator), do: accumulator
  defp factorial_iter(n, accumulator), do: factorial_iter(n - 1, n * accumulator)

  def sum_numbers(n), do: sum_numbers_iter(n, 0)

  defp sum_numbers_iter(0, accumulator), do: accumulator
  defp sum_numbers_iter(n, accumulator), do: sum_numbers_iter(n - 1, n + accumulator)
end
